Port aggregate counter sink

ilayaperumalg opened the pull request:
- Add aggregate counter sink configuration/properties
  - Add aggregate counter repository (redis/memory based)
  - Add tests

This resolves #102

ilayaperumalg committed:
Port aggregate counter sink

 - Add aggregate counter sink configuration/properties
 - Add aggregate counter repository (redis/memory based)
 - Add tests

This resolves #102

ericbottard commented on aggregate-counter-sink/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounterSink.java:
```diff
@@ -0,0 +1,75 @@
+
+	@ServiceActivator(inputChannel = Sink.INPUT)
+	public void process(Message<?> message) {
+		Double increment = this.config.getIncrementExpression().getValue(message, Double.class);
```
Why `Double` if we're gonna use `long` on line 57?

ericbottard commented on aggregate-counter-sink/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounterSink.java:
```diff
@@ -0,0 +1,75 @@
+		String counterName = this.config.getComputedNameExpression().getValue(message, CharSequence.class).toString();
+
+		if (this.config.getTimeField() == null) {
+			this.aggregateCounterRepository.increment(counterName, increment.longValue());
```
I would reduce the repository API surface and pass `now()` as a third parameter here

ericbottard commented on aggregate-counter-sink/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounterSink.java:
```diff
@@ -0,0 +1,75 @@
+		}
+		else {
+			String timeStampValue = this.config.getTimeField().getValue(message, String.class);
+			if (dateTimeFormatter == null) {
```
Why use a field here? Moreover, there could be a race condition here

artembilan replied:
:+1: 
if `DateTimeFormatter` isn't Thread Safe, should be better to use a new instance for each invocation, or `synchronized` around that `parseDateTime`.

From other side do we really need JodaTime for such a task ?

ericbottard commented on aggregate-counter-sink/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounterSinkProperties.java:
```diff
@@ -0,0 +1,126 @@
+	private Expression timeField;
+
+	/**
+	 *  Increment value for each bucket as a SpEL against the message
```
Additional space at start of javadoc and missing end dot.

ericbottard commented on aggregate-counter-sink/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounterSinkProperties.java:
```diff
@@ -0,0 +1,126 @@
+	}
+
+	public Expression getComputedNameExpression() {
+		return (nameExpression != null ? nameExpression : new LiteralExpression(getName()));
```
Seems weird that `nameExpression` is also tested in `getName()`

ilayaperumalg replied:
The `nameExpression` check in `getName()` is needed for the `isExclusiveOptions` assertion.

ericbottard commented on aggregate-counter-sink/src/test/java/org/springframework/cloud/stream/module/AggregateCounterTests.java:
```diff
@@ -0,0 +1,132 @@
+ */
+@RunWith(SpringJUnit4ClassRunner.class)
+@SpringApplicationConfiguration(classes = AggregateCounterSinkApplication.class)
+@DirtiesContext
```
Does it, really? Given that we reset the repository, seems like we can re-user the AppContext

ericbottard commented on aggregate-counter-sink/src/test/java/org/springframework/cloud/stream/module/AggregateCounterTests.java:
```diff
@@ -0,0 +1,132 @@
+	@Rule
+	public RedisTestSupport redisTestSupport = new RedisTestSupport();
+
+	private static final String AGGREGATE_COUNTER_NAME = "aggregate-counter-test.foo";
```
Not clear what the difference between these two names is.

ericbottard commented on common/spring-cloud-stream-modules-analytics/pom.xml:
```diff
@@ -32,5 +32,10 @@
+		<dependency>
+			<groupId>joda-time</groupId>
+			<artifactId>joda-time</artifactId>
+			<version>2.9.2</version>
```
Doesn't boot provide a version somehow?

ericbottard commented on common/spring-cloud-stream-modules-analytics/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounter.java:
```diff
@@ -0,0 +1,82 @@
+	@Override
+	public String toString() {
+		return "AggregateCount{" +
+				"name='" + name + '\'' +
```
Can put the closing single quote as part of the subsequent String

ericbottard commented on common/spring-cloud-stream-modules-analytics/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounterReader.java:
```diff
@@ -0,0 +1,63 @@
+
+	/**
+	 * Query function to allow the counts for a specific interval to be retrieved for the given resolution.
+	 *
```
extra line

ericbottard commented on time-source/src/main/java/org/springframework/cloud/stream/module/time/TimeSourceProperties.java:
```diff
@@ -17,6 +17,7 @@
 package org.springframework.cloud.stream.module.time;
 
 import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.cloud.stream.module.DateFormat;
```
import added but not used?

artembilan commented on aggregate-counter-sink/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounterSinkProperties.java:
```diff
@@ -0,0 +1,126 @@
+	/**
+	 *  Increment value for each bucket as a SpEL against the message
+	 */
+	private Expression incrementExpression = new SpelExpressionParser().parseExpression("1");
```
A `new ValueExpression<>(1);` would be much better.
I guess in most cases this expression will be used by default.

artembilan commented on aggregate-counter-sink/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounterSink.java:
```diff
@@ -0,0 +1,75 @@
+	@ServiceActivator(inputChannel = Sink.INPUT)
+	public void process(Message<?> message) {
+		Double increment = this.config.getIncrementExpression().getValue(message, Double.class);
+		String counterName = this.config.getComputedNameExpression().getValue(message, CharSequence.class).toString();
```
What is the point to use `CharSequence` and then convert it into `String` anyway?

ericbottard commented:
We need to review the while TimeZone handling of this (cf issues we encountered in the SmartGrid demo).

ilayaperumalg committed:
Address review comments

ilayaperumalg commented:
Rebased and updated the PR after addressing the review comments.

ilayaperumalg commented:
Created PR to update the `spring-cloud-dataflow` artifact registry populator: https://github.com/spring-cloud/spring-cloud-dataflow/pull/461

ericbottard commented:
A couple of issues:

In effect, the default store is not redis. This may be an issue with other counters after all, but we must find a working solution for this. Moreover, using `@ConditionalOnProperty` with a hardcoded property is likely to cause problems when we introduce a prefix. It would be better to rely on the actual instance of `AggregateCounterSinkProperties` somehow.

When actually using redis (explicit `--store=redis`), these are the keys I get when testing. Note the extra dot in there:
```
127.0.0.1:6379> keys *
 1) "spring.cloud.artifact.task"
 2) "aggregate-counters..aggregate-counter.20160321"
 3) "spring.cloud.artifact.sink"
 4) "aggregate-counters.aggregate-counter"
 5) "aggregate-counters..aggregate-counter.2016"
 6) "aggregate-counters..aggregate-counter.2016032114"
 7) "metric_meta.aggregatecounters.aggregate-counter"
 8) "groups.foo.time"
 9) "spring.cloud.artifact.processor"
10) "aggregate-counters..aggregate-counter.years"
11) "groups.foo.http"
12) "aggregate-counters..aggregate-counter.201603"
```

ericbottard commented on aggregate-counter-sink/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounterSinkProperties.java:
```diff
@@ -0,0 +1,128 @@
+
+	private String dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'";
+
+	private DateTimeFormatter dateFormatter;
```
unused

ilayaperumalg committed:
Avoid using ConditionalOnProperty

 - Use configuration properties to instantiate the appropriate repositories
  - This assumes the `spring-data-redis` is on the classpath
 - Apply the changes to counter, field-value-counter repository instantiations as well
 - Fix the repo prefix for Aggregate counter

ilayaperumalg commented:
Pushed the changes that avoids using ConditionalOnProperty:

 - Use configuration properties to instantiate the appropriate repositories
  - This assumes the `spring-data-redis` is on the classpath
 - Apply the changes to counter, field-value-counter repository instantiations as well
 - Fix the repo prefix for Aggregate counter

ericbottard commented:
Rebased, squashed as b378da5
